### PR TITLE
add AdaRanker model and test file

### PR DESCRIPTION
**Pull request by Master-PLC**
This is a PR for manually adding the AdaRanker model and its corresponding test files to the current UniRec codebase.

### CODE REVIEW

File: unirec/model/rank/adaranker.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                self.embedding_size, self.embedding_size, self.embedding_size, 
+                self.dropout_prob, self.device
+            )
+
+            # add bias layers
+            self.film_affine_emb_scale = nn.Linear(self.embedding_size, 1)
+            self.film_affine_emb_bias = nn.Linear(self.embedding_size, 1)
+
+            # predict_layer
+            self.mem_w1 = modules.MemoryUnit(self.dnn_input_size, self.dnn_inner_size, self.embedding_size)
```
Comment (Leavingseason) [inline unirec/model/rank/adaranker.py:42]: why hardcode 2 memory units?

**Review by Leavingseason — CHANGES_REQUESTED** [review]

File: examples/preprocess/specific_datasets/ranker.py [hunk 2]
```diff
@@ -325,7 +328,7 @@ def write_to_file(wt, user, items, labels=None):
         write_to_file(wt_valid, user, valid_items, labels)
         write_to_file(wt_test, user, test_items, labels)
 
-        write_to_file(wt_user_history, user, hist)
+        write_to_file(wt_user_history, user, hist[:-1])
```
[the remark this answers is not in the document]
Comment (Master-PLC) [reply examples/preprocess/specific_datasets/ranker.py:331]: this means remove the last item from user's interactive history, which is used as test target item.

File: examples/preprocess/specific_datasets/ranker.py [hunk 3]
```diff
@@ -336,6 +339,56 @@ def write_to_file(wt, user, items, labels=None):
...
+    print(user2itemseq)
+
+    corpus, item_num = load_item_seq(user2itemseq)
+    
+    vector_size = arguments['embedding_size']
+    w2v_model = Word2Vec(corpus, vector_size=vector_size, window=10, min_count=3, workers=4)
+    w2v_model.save(outfile_path+"/word2vec.model")
+
+
+    fp = open(outfile_path+'/item_emb_'+str(vector_size)+'.txt', 'w')
```
[the remark this answers is not in the document]
Comment (Master-PLC) [reply examples/preprocess/specific_datasets/ranker.py:381]: ok, I'll check the way to generate a file or directory path.

File: examples/preprocess/specific_datasets/ranker.py [hunk 4]
```diff
@@ -336,6 +339,56 @@ def write_to_file(wt, user, items, labels=None):
...
+    for i in tqdm(range(1, item_num+1), desc='output Item Emb'):
```
[the remark this answers is not in the document]
Comment (Master-PLC) [reply examples/preprocess/specific_datasets/ranker.py:382]: cause itemid in corpus starts from 1

File: tests/test_dataset/test_preprocess.py [hunk 5]
```diff
@@ -210,4 +215,4 @@ def test_preprocess_rank_data():
 
 
 if __name__ == "__main__":
-    pytest.main(["test_morec.py", "-s"])
```
[the remark this answers is not in the document]
Comment (Master-PLC) [reply tests/test_dataset/test_preprocess.py:213]: I made a mistake that I thought this was wrongly wrote as test_morec.py instead of test_preprocess.py. I'll roll back it.

File: unirec/facility/trainer.py [hunk 6]
```diff
@@ -377,6 +377,14 @@ def load_model(self, filename=None):
         unwrapped_model = self.accelerator.unwrap_model(self.model)
         unwrapped_model.load_state_dict(checkpoint['state_dict'], strict=False)
         self.logger.info('Loading model from {0}. The best epoch was {1}'.format(checkpoint_file, checkpoint['cur_epoch']))
+        ### freeze base model's parameters
+        if self.config['freeze']:
```
[the remark this answers is not in the document]
Comment (Master-PLC) [reply unirec/facility/trainer.py:381]: ok, though I think this can be used at other place

File: unirec/facility/trainer.py [hunk 7]
```diff
@@ -377,6 +377,14 @@ def load_model(self, filename=None):
...
+            self.logger.info('Freeze the pretrained model parameters.')
+            for name, param in unwrapped_model.named_parameters():
+                if name in checkpoint['state_dict']:
```
[the remark this answers is not in the document]
Comment (Master-PLC) [reply unirec/facility/trainer.py:384]: no, when we use AdaRanker to finetune the base model such as GRU, this step only freeze the parameters in AdaRanker which have the same name in GRU, the plug-and-play parameters in AdaRankers still can be trained.

File: unirec/model/rank/adaranker.py (hunk 1 above)
Comment (Master-PLC) [reply unirec/model/rank/adaranker.py:42]: cause self.mlp_layers is a two-layer MLP, we need 2 memory units for each layer as w_add and b_add.

File: unirec/utils/argument_parser.py [hunk 8]
```diff
@@ -147,6 +147,11 @@ def parse_cmd_arguments():
     ## FM model
     parser.add_argument("--linear_mode", type=str, help='sparse linear mode in FM')
 
+    ## AdaRanker model
+    parser.add_argument("--train_type", type=str, default="Base", help='train type for AdaRanker')
```
Comment (XuHwang) [inline unirec/utils/argument_parser.py:151]: maybe we could use `choices` for the argument to indicate what train types we support.

Comment (XuHwang) [reply unirec/utils/argument_parser.py:151]: It is better to add those added arguments into `base.yaml` with default values.

File: unirec/model/modules.py [hunk 9]
```diff
@@ -430,3 +431,280 @@ def forward(self, hidden_states, attention_mask, output_all_encoded_layers=True)
         if not output_all_encoded_layers:
             all_encoder_layers.append(hidden_states)
         return all_encoder_layers
+
+
+# Ada-Ranker
+class NeuProcessEncoder(nn.Module):
```
Comment (XuHwang) [inline unirec/model/modules.py:437]: it is suggested to add some annotations for the new modules.

File: unirec/model/rank/adaranker.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
+import torch
+import torch.nn as nn
+
+from ..base.ranker import Ranker
+import unirec.model.modules as modules
+
+
+class AdaRanker(Ranker):
```
[the remark this answers is not in the document]
Comment (Master-PLC) [reply unirec/model/rank/adaranker.py:8]: ok, I've added some descriptions to AdaRanker.

File: unirec/utils/argument_parser.py (hunk 8 above)
Comment (Master-PLC) [reply unirec/utils/argument_parser.py:151]: sounds great!

File: unirec/model/modules.py (hunk 9 above)
Comment (Master-PLC) [reply unirec/model/modules.py:437]: Got it!

File: unirec/model/rank/adaranker.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            subsequent_mask = subsequent_mask.long().to(item_seq.device)
+
+            extended_attention_mask = extended_attention_mask * subsequent_mask
+
+        extended_attention_mask = extended_attention_mask.to(dtype=next(self.parameters()).dtype)  # fp16 compatibility
+        extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
+        return extended_attention_mask
+
+    def forward_user_emb(self, item_seq=None):
+        item_seq_emb = self.item_embedding(item_seq)
```
[the remark this answers is not in the document]
Comment (Master-PLC) [reply unirec/model/rank/adaranker.py:114]: 因为AdaRanker目前还没有这么复杂的需求，暂时不考虑太复杂的设计

[the remark this answers is not in the document]
Comment (Leavingseason) [reply unirec/model/rank/adaranker.py:114]: 需要写个TODO, 标注一下目前的版本暂不支持这些功能

File: examples/preprocess/download_split_ml100k.py [hunk 12]
```diff
@@ -148,6 +148,41 @@ def k_core_filter(df: pd.DataFrame, user_k=10, item_k=10, user_col_name='user_id
     return df
 
 
+
+def merge_category(data):
+    # Get cate2items and item2cate
+    # Merge categories containing a small number of items (lower than 200) into one category
```
Comment (Leavingseason) [inline examples/preprocess/download_split_ml100k.py:154]: 数字200对不上。

Comment (Leavingseason) [reply examples/preprocess/download_split_ml100k.py:154]: 描述一下input 和 output的格式

File: examples/preprocess/run_prepare_data-ml-100k-rank.sh [hunk 13]
```diff
@@ -1,9 +1,11 @@
...
-RAW_DATA_PREFILE="/path/to//data/ml-100k/user_history.csv"
-RAW_DATA_FILE="/path/to/raw_datasets/ml-100k-rank/ml-100k-rank.txt"
-RAW_DATA_DIR="/path/to/raw_datasets"
+HOME_DIR=$(eval echo ~)
+
+RAW_DATA_DIR="$HOME_DIR/.unirec/dataset"
+RAW_DATA_PREFILE="$RAW_DATA_DIR/ml-100k/full_user_history.csv"
+RAW_DATA_FILE="$RAW_DATA_DIR/ml-100k-rank/ml-100k-rank.txt"
 
 ROOT_DIR='/path/to/UniRec'
```
Comment (Leavingseason) [inline examples/preprocess/run_prepare_data-ml-100k-rank.sh:10]: 尽量把用户需要根据实际情况手动修改的地方放在开头， 并用符号标记出来，比如：
##########################################################################
######## Please modify the following variables according to your device                       ########
##########################################################################

ROOT_DIR='/path/to/UniRec'

##########################################################################

File: examples/preprocess/specific_datasets/ranker.py [hunk 14]
```diff
@@ -325,7 +329,7 @@ def write_to_file(wt, user, items, labels=None):
         write_to_file(wt_valid, user, valid_items, labels)
         write_to_file(wt_test, user, test_items, labels)
 
-        write_to_file(wt_user_history, user, hist)
+        write_to_file(wt_user_history, user, hist[:-1])
```
Comment (Leavingseason) [inline examples/preprocess/specific_datasets/ranker.py:332]: 建议还原为hist

File: examples/preprocess/specific_datasets/ranker.py [hunk 15]
```diff
@@ -336,15 +340,246 @@ def write_to_file(wt, user, items, labels=None):
     print(f'min item length: {min(lengths)}')
 
 
+def sample_while(candi_items, sampled_items_num, target_item, item_seq):
```
Comment (Leavingseason) [inline examples/preprocess/specific_datasets/ranker.py:343]: 函数名得换一个专业的

File: examples/preprocess/specific_datasets/ranker.py [hunk 16]
```diff
@@ -336,15 +340,246 @@ def write_to_file(wt, user, items, labels=None):
...
+    cur_sampled_items = []
+    while len(cur_sampled_items) < sampled_items_num:
+        sampled_item = random.sample(candi_items, 1)[0]
+        while (sampled_item == target_item) or (sampled_item in item_seq) or (sampled_item in cur_sampled_items):
```
Comment (Leavingseason) [inline examples/preprocess/specific_datasets/ranker.py:347]: 需要加一个max_retry_cnt, 避免有些极端情况陷入死循环

File: examples/preprocess/specific_datasets/ranker.py [hunk 17]
```diff
@@ -336,15 +340,246 @@ def write_to_file(wt, user, items, labels=None):
...
+    sampled_items = []
+
+    seed = random.sample(range(0, 100), 1)[0]
+    for idx in range(len(sample_num)):
+        sampled_items_num = sample_num[idx]
+        if sampled_items_num == 0:
+            continue
+        cate = sampled_cate[idx]
+        # 50% universal 50% pop
+        if seed < 100: # universal
```
Comment (Leavingseason) [inline examples/preprocess/specific_datasets/ranker.py:371]: seed = random.sample(range(0, 100), 1)[0]  所以seed是不是总小于100？

File: examples/preprocess/specific_datasets/ranker.py [hunk 18]
```diff
@@ -336,15 +340,246 @@ def write_to_file(wt, user, items, labels=None):
...
+        cur_sampled_items = sample_while(candi_items, sampled_items_num, tgt_item, item_to_remove)
+        sampled_items += cur_sampled_items
+
+    return sampled_items
+
+
+r'''
+This function preprocesses the user history file and generates various output files for training, validation, and testing in T4 data format.
+
+Args:
```
Comment (Leavingseason) [inline examples/preprocess/specific_datasets/ranker.py:385]: 这个注释写得挺好

File: tests/test_dataset/test_preprocess.py [hunk 19]
```diff
@@ -57,105 +57,174 @@
...
+#     assert os.path.exists(os.path.join(pre_config['outdir'], 'train.txt')), "raw train file not created sucessfully"
+#     assert os.path.exists(os.path.join(pre_config['outdir'], 'valid.txt')), "raw valid file not created sucessfully"
+#     assert os.path.exists(os.path.join(pre_config['outdir'], 'test.txt')), "raw test file not created sucessfully"
+#     assert os.path.exists(os.path.join(pre_config['outdir'], 'user_history.txt')), "raw user_history file not created sucessfully"
+#     assert os.path.exists(os.path.join(pre_config['outdir'], 'user2uid.txt')), "user to user_id file not created sucessfully"
+#     assert os.path.exists(os.path.join(pre_config['outdir'], 'item2tid.txt')), "item to target_id file not created sucessfully"
+#     assert os.path.exists(os.path.join(pre_config['outdir'], 'word2vec.model')), "word2vec model file not created sucessfully"
+#     assert os.path.exists(os.path.join(pre_config['outdir'], f"item_emb_{pre_config['embedding_size']}.txt")), "pretrained item embedding file not created sucessfully"
+
+#     # processed data file could be used for further test pipeline
```
Comment (Leavingseason) [inline tests/test_dataset/test_preprocess.py:176]: 为什么有大段的代码被注释掉？

**Review by Leavingseason — CHANGES_REQUESTED** [review]

File: examples/preprocess/download_split_ml100k.py (hunk 12 above)
Comment (Master-PLC) [reply examples/preprocess/download_split_ml100k.py:154]: 改成一个可以调整的值了

File: examples/preprocess/run_prepare_data-ml-100k-rank.sh (hunk 13 above)
Comment (Master-PLC) [reply examples/preprocess/run_prepare_data-ml-100k-rank.sh:10]: 好的，需要吧大部分的example都调整一下吗

File: examples/preprocess/specific_datasets/ranker.py (hunk 14 above)
Comment (Master-PLC) [reply examples/preprocess/specific_datasets/ranker.py:332]: 好的

File: examples/preprocess/specific_datasets/ranker.py (hunk 15 above)
Comment (Master-PLC) [reply examples/preprocess/specific_datasets/ranker.py:343]: 改成sample_items_from_candidates

File: tests/test_dataset/test_preprocess.py (hunk 19 above)
Comment (Master-PLC) [reply tests/test_dataset/test_preprocess.py:176]: 还是漏检查了，我再check一下

File: examples/preprocess/specific_datasets/ranker.py (hunk 17 above)
Comment (Master-PLC) [reply examples/preprocess/specific_datasets/ranker.py:371]: 是的，之前测试的时候设置的小于100，现在重新采用了50% popularity和50% universa

File: examples/preprocess/specific_datasets/ranker.py (hunk 16 above)
Comment (Master-PLC) [reply examples/preprocess/specific_datasets/ranker.py:347]: 去除了while sample的机制，通过assert先判断是否能从candi_items中sample出来不在hist中的k个负样本

File: examples/preprocess/run_prepare_data-ml-100k-rank.sh (hunk 13 above)
Comment (Leavingseason) [reply examples/preprocess/run_prepare_data-ml-100k-rank.sh:10]: 好的
